File: Backtrack (Recursion)/N-Queen.cpp
#include<bits/stdc++.h>
using namespace std; 

#define int long long

int cnt = 0;
int a[11];
int n;

bool check(int i, int j) {
    //check if we can place a queen at (i, j) not attacking any other queens from 1 to i-1
    for(int k = 1; k < i; k++) {
        if(a[k] == j || abs(a[k] - j) == abs(k - i)) return false;
    }
    return true;
}

int backtrack(int i) {
    int cnt = 0;
    for(int j = 1; j <= n; j++) {
        if(check(i, j)) {
            a[i] = j;
            if(i == n) {
                cnt++;
            }
            else {
                cnt += backtrack(i+1);
            }
        }
    }
    return cnt;
}

signed main() {
    cin >> n;
    cout << backtrack(1);
    return 0;
}

File: Backtrack (Recursion)/N-Queen.py
import math

n = int(input())
a = [0 for i in range(n+1)]

def check(i, j):
    for k in range(1, i):
        if a[k] == j or abs(a[k] - j) == abs(k-i):
            return False
    return True

def backtrack(i) -> int:
    cnt = 0
    for j in range(1, n+1):
        if check(i, j):
            a[i] = j
            if i == n:
                cnt += 1
            else:
                cnt += backtrack(i+1)
    return cnt

print(backtrack(1))